test/testtokenize.cpp

```diff
@@ -7904,6 +7966,25 @@ class TestTokenizer : public TestFixture {
         ASSERT_EQUALS("b(", testAst("class a { void b() & {} };"));
     }
 
+    void astnewscoped() {
+        auto generateAst = [&](std::string code) {
+            Settings settings;
+            Library library;
+            auto tokenizer = std::make_unique<Tokenizer>(&settings, this);
+            std::istringstream istr(code);
+            tokenizer->tokenize(istr, "test.cpp");
+            return tokenizer;
+        };
+
+        //tokenizer.
+        bool match1 = SimpleAst{"return", {"new", {"A"}}}.matches(*generateAst("return new A;"));
+        bool match2 = SimpleAst{"return", {"new", {"(", {"A"}}}}.matches(*generateAst("return new A();"));
+        bool match3 = SimpleAst{"return", {"new", {"(", {"A"}, {"true"}}} }.matches(*generateAst("return new A(true);"));
+        bool match4 = SimpleAst{"return", {"new", {"::", {"A"}, {"B"}}}}.matches(*generateAst("return new A::B;"));
+        bool match5 = SimpleAst{"return", {"new", {"(", {"::", {"A"}, {"B"}}}}}.matches(*generateAst("return new A::B();"));
+        bool match6 = SimpleAst{"return", {"new", {"(", {"::", {"A"}, {"B"}}, {"true"}}}}.matches(*generateAst("return new A::B(true);"));
```

Comment:
My intention was to go all the way up to ::D but it was getting late and I wanted to get a rough outline of the change up for people to look at in case I was way off. You're right though ::C is an important test case which would have caught the shadow variable bug you spotted.